File: euler35.py
# Сколько существует круговых простых чисел меньше миллиона?

from datetime import datetime
from typing import List


def eratosfen(number: int) -> List[int]:
    """
    Решето Эратосфена: генерация простых чисел вплоть до заданного числа.

    Args:
        number: Верхняя граница диапазона.

    Returns:
        Список, где индекс - число, значение - 0 для составных,
        само число для простых.
    """
    sieve = list(range(number + 1))
    sieve[1] = 0
    for i in sieve:
        if i > 1:
            for j in range(2 * i, len(sieve), i):
                sieve[j] = 0
    return sieve


def euler35() -> None:
    """
    Решение задачи Эйлера №35.

    Сколько существует круговых простых чисел меньше миллиона?
    """
    n = 1_000_000

    # 1 вариант
    start_time = datetime.now()
    primes, circular_primes = set(eratosfen(n)) - {0}, []
    for i in primes:
        if len(set(str(i))) == 1:
            circular_primes.append(i)
            continue

        next_num, circular_primes_check, power = i, [i], len(str(i)) - 1
        for _ in range(power):
            next_num = next_num % 10 * 10 ** power + next_num // 10
            if next_num in primes:
                circular_primes_check.append(next_num)
            else:
                circular_primes_check = []
                break
        circular_primes += [
            x for x in circular_primes_check if x not in circular_primes
        ]

    print(len(circular_primes))
    print(datetime.now() - start_time)

    # 2 вариант
    start_time = datetime.now()
    primes_bool = [True] * (n + 1)
    i = 2
    while i * i <= n:
        if primes_bool[i]:
            for j in range(i * 2, n + 1, i):
                primes_bool[j] = False
        i += 1

    primes_bool[0] = primes_bool[1] = False
    count = 0
    for i in range(1, n + 1):
        if primes_bool[i] is False:
            continue
        next_num, flag = str(i), True
        for _ in next_num:
            next_num = next_num[1:] + next_num[0]
            flag &= primes_bool[int(next_num)]
        if flag:
            count += 1
    print(count)
    print(datetime.now() - start_time)


if __name__ == "__main__":
    euler35()

File: go/euler35.go
// euler35 считает количество круговых простых чисел меньше миллиона.
//
// Решение задачи Эйлера №35.
package main

import (
	"fmt"
	"time"
)

func eratosfen(number int) []bool {
	sieve := make([]bool, number+1)
	for i := range sieve {
		sieve[i] = true
	}
	sieve[0] = false
	sieve[1] = false
	for i := 2; i*i <= number; i++ {
		if sieve[i] {
			for j := i * i; j <= number; j += i {
				sieve[j] = false
			}
		}
	}
	return sieve
}

func rotate(n int, power int) int {
	s := fmt.Sprintf("%d", n)
	rotated := s[1:] + s[:1]
	result := 0
	fmt.Sscan(rotated, &result)
	return result
}

func euler35() {
	n := 1000000

	startTime := time.Now()
	sieve := eratosfen(n)
	primes := make(map[int]bool)
	for i := 2; i <= n; i++ {
		if sieve[i] {
			primes[i] = true
		}
	}
	delete(primes, 0)
	delete(primes, 1)
	circularPrimes := make([]int, 0)
	for i := range primes {
		hasZero := false
		for _, c := range fmt.Sprintf("%d", i) {
			if c == '0' {
				hasZero = true
				break
			}
		}
		if hasZero {
			continue
		}

		power := len(fmt.Sprintf("%d", i)) - 1
		nextNum := i
		circularPrimesCheck := []int{i}
		for j := 0; j < power; j++ {
			nextNum = rotate(nextNum, power)
			if primes[nextNum] {
				circularPrimesCheck = append(circularPrimesCheck, nextNum)
			} else {
				circularPrimesCheck = nil
				break
			}
		}
		if circularPrimesCheck != nil {
			for _, x := range circularPrimesCheck {
				if !contains(circularPrimes, x) {
					circularPrimes = append(circularPrimes, x)
				}
			}
		}
	}
	fmt.Println(len(circularPrimes))
	fmt.Println(time.Since(startTime))

	startTime = time.Now()
	primesBool := make([]bool, n+1)
	for i := range primesBool {
		primesBool[i] = true
	}
	i := 2
	for i*i <= n {
		if primesBool[i] {
			for j := i * 2; j <= n; j += i {
				primesBool[j] = false
			}
		}
		i++
	}
	primesBool[0] = false
	primesBool[1] = false
	count := 0
	for i = 1; i <= n; i++ {
		if !primesBool[i] {
			continue
		}
		nextNum := fmt.Sprintf("%d", i)
		flag := true
		for range nextNum {
			nextNum = nextNum[1:] + nextNum[:1]
			num := 0
			fmt.Sscan(nextNum, &num)
			flag = flag && primesBool[num]
		}
		if flag {
			count++
		}
	}
	fmt.Println(count)
	fmt.Println(time.Since(startTime))
}

func contains(slice []int, val int) bool {
	for _, v := range slice {
		if v == val {
			return true
		}
	}
	return false
}

func main() {
	euler35()
}
